epaint/src/text/text_layout.rs

```diff
@@ -170,18 +180,28 @@ fn rows_from_paragraphs(paragraphs: Vec<Paragraph>, wrap_width: f32) -> Vec<Row>
     rows
 }
 
-fn line_break(paragraph: &Paragraph, wrap_width: f32, out_rows: &mut Vec<Row>) {
+fn line_break(
+    fonts: &mut FontsImpl,
+    paragraph: &Paragraph,
+    job: &LayoutJob,
+    out_rows: &mut Vec<Row>,
+) {
     // Keeps track of good places to insert row break if we exceed `wrap_width`.
     let mut row_break_candidates = RowBreakCandidates::default();
 
     let mut first_row_indentation = paragraph.glyphs[0].pos.x;
     let mut row_start_x = 0.0;
     let mut row_start_idx = 0;
+    let mut non_empty_rows = 0;
 
     for (i, glyph) in paragraph.glyphs.iter().enumerate() {
         let potential_row_width = glyph.max_x() - row_start_x;
 
-        if potential_row_width > wrap_width {
+        if job.wrap.max_lines > 0 && non_empty_rows >= job.wrap.max_lines {
+            break;
+        }
+
+        if potential_row_width > job.wrap.max_width {
             if first_row_indentation > 0.0 && !row_break_candidates.has_word_boundary() {
```

Comment:
`break_anywhere` should be taken into account here, e.g. with `row_break_candidates.has_good_candidate(job.wrap.break_anywhere)`

Reply:
Sorry, I'm being a dum-dum, as in, `potential_row_width > job.wrap.max_width && row_break_candidates.get(job.wrap.break_anywhere).is_some()`? That seemingly breaks the text breaks.

![image](https://user-images.githubusercontent.com/20689137/157060714-a0a5ad19-ec0f-49da-bb0d-9d0779ae181c.png)

Reply:
I meant we should add something like

``` rust
impl RowBreakCandidates {
    fn has_good_candidate(&self, break_anywhere: bool) -> bool {
        if break_anywhere { self.any.is_some() }
        else { self.has_word_boundary() }
    }
}
```